rust/connlib/connection/src/pool.rs

```diff
@@ -0,0 +1,632 @@
+use pnet_packet::Packet;
+use rand::random;
+use secrecy::{ExposeSecret, Secret};
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::time::{Duration, Instant};
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    net::SocketAddr,
+    sync::Arc,
+};
+use str0m::ice::{IceAgent, IceAgentEvent, IceCreds};
+use str0m::net::{Protocol, Receive};
+use str0m::{Candidate, StunMessage};
+
+use crate::index::IndexLfsr;
+use crate::IpPacket;
+
+// Note: Taken from boringtun
+const HANDSHAKE_RATE_LIMIT: u64 = 100;
+
+const MAX_UDP_SIZE: usize = (1 << 16) - 1;
+
+/// Manages a set of wireguard connections for a server.
+pub type ServerConnectionPool<TId> = ConnectionPool<Server, TId>;
+/// Manages a set of wireguard connections for a client.
+pub type ClientConnectionPool<TId> = ConnectionPool<Client, TId>;
+
+pub enum Server {}
+pub enum Client {}
+
+pub struct ConnectionPool<T, TId> {
+    private_key: StaticSecret,
+    index: IndexLfsr,
+    rate_limiter: Arc<RateLimiter>,
+    local_interfaces: HashSet<SocketAddr>,
+    buffered_transmits: VecDeque<Transmit>,
+
+    next_rate_limiter_reset: Option<Instant>,
+
+    initial_connections: HashMap<TId, InitialConnection>,
+    negotiated_connections: HashMap<TId, Connection>,
+    pending_events: VecDeque<Event<TId>>,
+
+    buffer: Box<[u8; MAX_UDP_SIZE]>,
+
+    marker: PhantomData<T>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Unknown interface")]
+    UnknownInterface,
+    #[error("Failed to decapsulate: {0:?}")] // TODO: Upstream an std::error::Error impl
+    Decapsulate(boringtun::noise::errors::WireGuardError),
+    #[error("Failed to encapsulate: {0:?}")]
+    Encapsulate(boringtun::noise::errors::WireGuardError),
+    #[error("Unmatched packet")]
+    UnmatchedPacket,
+    #[error("Not connected")]
+    NotConnected,
+}
+
+impl<T, TId> ConnectionPool<T, TId>
+where
+    TId: Eq + Hash + Copy + fmt::Display,
+{
+    pub fn new(private_key: StaticSecret) -> Self {
+        let public_key = &(&private_key).into();
+        Self {
+            private_key,
+            marker: Default::default(),
+            index: IndexLfsr::default(),
+            rate_limiter: Arc::new(RateLimiter::new(public_key, HANDSHAKE_RATE_LIMIT)),
+            local_interfaces: HashSet::default(),
+            buffered_transmits: VecDeque::default(),
+            next_rate_limiter_reset: None,
+            negotiated_connections: HashMap::default(),
+            pending_events: VecDeque::default(),
+            initial_connections: HashMap::default(),
+            buffer: Box::new([0u8; MAX_UDP_SIZE]),
+        }
+    }
+
+    pub fn add_local_interface(&mut self, local_addr: SocketAddr) {
+        self.local_interfaces.insert(local_addr);
+
+        // TODO: Add host candidate to all existing connections here.
+    }
+
+    pub fn add_remote_candidate(&mut self, id: TId, candidate: String) {
+        let candidate = match Candidate::from_sdp_string(&candidate) {
```

Comment:
😢

Reply:
This looks a bit odd I know. My thinking was that I wanted to encapsulate the dependencies within `firezone-connection` as much as possible. Currently, `str0m` is a private dependency of it. `boringtun` leaks into the public API due to its key types and errors but it wouldn't be too hard to encapsulate that too.

Even though we aren't planning on releasing this to crates.io soon (ever?), I think we will benefit from containing dependencies as much as possible within certain libraries.

For candidates, there was a choice of either exposing `str0m`, using a `String` and SDP format or a custom new-type. To move faster, I opted for the `String` for now. Long-term, I think it would be cleaner to have our own new-type that encapsulates `str0m`'s candidate and parses early to avoid the error case you are pointing out here. But that was more effort than I wanted to spend here for now.

if you agree with the direction, I'll put it down as a TODO and fix at some point :)